.npmignore

```diff
@@ -1,44 +1,34 @@
-*.scssc
-*.zip
-*.gzip
-*.tgz
-*.DS_Store
-
-*.sass-cache
-*COMPILE.css
-
-# Intellij project configuration files
-*.idea
-*.iml
-
-# External dependencies
-
-# Build output
-target
-
-# Mac OS X Finder
-.DS_Store
-
-# Closed source libraries
-closed-lib
-
-# Node, Bower dependencies
-node_modules
-bower_components
-
-Procfile
-
-# Protractor logs
-protractor/logs
-
-# npm-debug log
-npm-debug.log
-
-# Infra and tests
-.circleci
-.github
-e2e
-codecov.yml
-lighthouserc.yml
-*.Spec.js
-karma.conf.js
+# Specificty of the following rules matters. If more specific rules comes first,
+# less specific rules can not override them.
+# For example, the following won't work as you may assume:
+# /**/* <-- ignore everything
+# !/src/**/* <-- don't ignore src/
+# But still ignore test files in src/:
+# /**/*.test.* <-- HERE: This won't work, is not as specific as `!/dist/**/*`, so test files inside dist or src will NOT be ignored.
+# /src/**/*.test.* <-- But this will work, because it is more specific.
+
+# NOTE! Although we ignored everything by default, NPM will not ignore special files like
+# LICENSE.md, README.md, and package.json.
+
+# Ignore everything,
+/**/*
+
+# ...but include these folders...
+!/dist/**/*
+!/src/**/*
+!/platform/**/*
+
+# ...except for these files in the above folders (these rules are more specific, so they will apply).
+/src/**/*Spec.js
+/src/**/test/
+# TODO move test utils into test/ folders
+/src/utils/testing.js
```

Comment:
Wdyt think about this TODO?

Reply:
There already is a `/src/**/test/` rule ignore anything in `test/` folders. It would make sense to put all non-spec test files like testing utils in such folders, to make it easier to ignore them.

`/src/utils/testing.js` was previously being published to npm.

We can also add one-off rules for specific test files to ignore, but I think it would be easier to have a convention that we just put any test files (apart from the specs that are co-located with the thing they test) in `test/` folders.

Reply:
@trusktr maybe we could file an issue for that chance and include a link to that issue in this comment?

Reply:
We have a naming convention of *Spec*. why can't we use that?

Reply:
@shefalijoshi Yes, true that, but multiple spec files like to import test-specific utilities from `src/utils/testing.js` so I had to specifically ignore that file because it isn't following a catch-all convention. So what I meant above is that instead we could move that file to `src/test/utils.js` instead, and that way it will fall under the `src/**/test/*` catch-all instead of having a one-off entry in npmignore.

Reply:
Hmm. Also, for ignoring everything, won't we need to keep updating this when we add new folders?
I prefer ignoring specific files to excluding everything and then including specific items.

Reply:
> won't we need to keep updating this when we add new folders?

What folders do you imagine we might add? I think keeping things minimal -- src for source and dist for all output -- is the way to go. I believe even the `platform/` folder should go in `src/` because it is source code.

I think it is easier to understand what is included in the published lib (and keeping that simple) rather than having to understand what isn't part of the lib.

The `*Spec.js` lines are the only exception since they're mixed with source files, and they need to be ignore; plus I don't imagine we need to add other types of files into `src/`.

Reply:
The `.npmignore` that I simply copy/paste into all my JS/TS packages looks like this:

```conf
# ignore everything first (does not ignore special files like README.md, package.json, LICENSE.md, etc)
/**/*

# but include these folders
!/dist/**/*
!/src/**/*

# except for test files in those folders.
/dist/**/*.test.*
/dist/tests/**/*
/src/**/*.test.*
/src/tests/**/*
```

Then from project to project it is really easy to manage. Each project differs and may have different things to ignore, but by convention I can rely on `src/` and `dist/` folders (with test files in them) being the standard, and it stays consistent across projects.

Reply:
@shefalijoshi Here's a good example!

Recently we split `webpack.config.js` into three files, `webpack.common.js`, `webpack.dev.js`, and `webpack.prod.js`. With the ignore method, we would need to update `npmignore` on any such change (master branch currently does not ignore those files, as well as others, see unpkg.com). With the include method, we wouldn't have to touch `npmignore`.

This is better. I hope I can convince you. 😃

Reply:
If we do go with this approach, we should add `example` to this too. It has `eventGenerator` and `generator` which a lot of people use to demo OpenMCT

Reply:
> If we do go with this approach, we should add `example` to this too. It has `eventGenerator` and `generator` which a lot of people use to demo OpenMCT

@trusktr Please include the example, docs folders and Contributing.md, API.md, app.js as well.
What is the procfile used for?

Reply:
We want to publish all those things in the package on NPM? I figured people can go to the repo for that stuff. People don't normally look for that stuff in `node_modules`.

Reply:
Oh I see @scottbell, so some people will want to import from examples.

Reply:
Do people import `app.js` from `node_modules`. My understanding is we want to obliterate that ( says @akhenry )

Reply:
> What is the procfile used for?

https://www.google.com/search?q=procfile&oq=procfile&aqs=edge..69i57.870j0j1&sourceid=chrome&ie=UTF-8

Not sure why we need that. Are we using it on some service? If some downstream project needs it, they should have that in their repo instead.